ENCD-5614-cart-dataset-filters

forresttanaka opened the pull request:
* Instead of tracking only the `@ids` of datasets, and retrieving a page of dataset objects every time the user wants to view a new page of datasets, I now load and keep all datasets (just a subset of their properties) just as file objects have worked. So functions like `CartSearchResults` no longer have to fetch dataset objects.
* We now load dataset descriptions on the cart-view page so that the selected annotations appear with full descriptions, just as they do on the annotation search page.
* `FileFacets` no longer resets facets when switching between showing default files only and all files, so we no longer need the useUpdate callback.
* I have a lot of duplication between file-facet code and dataset-facet code. I looked into trying to consolidate some of that, but it doesn’t look straightforward. Maybe someday.

forresttanaka committed:
ENCD-5614-cart-dataset-filters

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -92,16 +93,47 @@ const analysisFieldMap = (analysisTitles) => (
 /**
- * List of allowed dataset types. Generally maps from collection names (e.g. '/experiments/' to a
- * displayable title.
+ * Field `getValue` function so that files without a `biosample_ontology` can have a "None" facet
```
This comment is okay. Still, "function" in code to me means "JavaScript object" rather than "purpose/role" as you intended. So it took me a second reading to understand. Also, since "getValue" is singular, "function" which is a verb, should be plural (functions).

forresttanaka replied:
I agree it reads awkwardly. Not sure if I improved it or not.

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -184,79 +217,96 @@ const displayedFacetFields = [
+const CartSearchResults = ({ elements, currentPage, cartControls }) => {
+    if (elements.length > 0) {
+        const pageStartIndex = currentPage * PAGE_ELEMENT_COUNT;
+        const currentPageElements = elements.slice(pageStartIndex, pageStartIndex + PAGE_ELEMENT_COUNT);
```
pageStartIndex + PAGE_ELEMENT_COUNT can exceed elements.length and there is no check for this. But JavaScript fails gracefully.

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -382,67 +432,53 @@ CartFiles.defaultProps = {
 
-        return <progress value={facetLoadProgress} max="100" />;
+    if (facetLoadProgress === -1) {
+        const fileCountFormatted = count.toLocaleString ? count.toLocaleString() : count.toString();
```
Good idea. Some numbers differ from America. Like in parts of Europe, "1,000" is written as "1.000"

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -1186,6 +1349,66 @@ CartPager.propTypes = {
+ * @return {array} `datasets` with search-result datasets added
+ */
+const addToAccumulatingDatasets = (datasets, currentResults) => {
+    if (currentResults['@graph'] && currentResults['@graph'].length > 0) {
```
Storing currentResults['@graph'] may make the code better, albeit computationally it won't make a difference since access to array entry is O(1)

zoldello commented on src/encoded/static/components/objectutils.js:
```diff
@@ -870,6 +870,20 @@ export const filterForDefaultFiles = (fileList) => (
+/**
+ * Filter the given files to only include those that appear in the given datasets.
+ * @param {array} fileList Files to filter
+ *
```
Why the gap?

forresttanaka replied:
There was something I saw that seemed authoritative that there should be a gap before @returns, but now I don’t see it and wonder if I imagined it. I’ll not do that anymore.

zoldello reviewed (dismissed):
Looks okay and UI looks okay too. My comments on the PR have nothing significant that needs changing.

forresttanaka commented:
I’m going to have to rebase to get a demo started.

forresttanaka committed:
ENCD-5614-cart-dataset-filters

Fixed a React warning from a missing property, and fixed up the cart BDD tests for a merge conflict.

zoldello commented on src/encoded/tests/features/cart.feature:
```diff
@@ -8,7 +8,7 @@ Feature: Cart
         And I press "/experiments/ENCSR000INT/"
         And I press "/experiments/ENCSR000DZQ/"
-        And I wait for the content to load
+        And I wait for 5 seconds
```
Why did you use 5 seconds rather than wait until the DOM-object is loaded?

forresttanaka replied:
I found that waiting for the content to load didn’t help — the test just barged on even if the cart was still loading.

forresttanaka committed:
ENCD-5614-cart-dataset-filters

Base all tab file lists from files from the selected datasets instead of all files.

forresttanaka committed:
ENCD-5614-cart-dataset-filters

Implement facet load progress on the dataset facet.

forresttanaka committed:
ENCD-5614-cart-dataset-filters

Correct a comment.

forresttanaka commented:
Redid how the raw and genome browser file lists get calculated so that they take the currently selected set of datasets into account. I also enabled the progress bar in the dataset facet as I had originally thought it wasn’t necessary, but using it with large carts proved me wrong.

Updated demo: https://encd-5614-cart-dataset-filters-fytanaka.demo.encodedcc.org/

I’ll start a new clean demo for tomorrow.

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -1113,6 +1114,8 @@ DatasetFacets.propTypes = {
     /** Callback when the user clicks on a file format facet item */
     termClickHandler: PropTypes.func.isRequired,
+    /** Facet-loading progress for progress bar, or null if not displayed */
+    facetLoadProgress: PropTypes.number,
```
Should you use PropTypes.oneOfType instead of PropTypes.number to cover the null case?

zoldello reviewed (dismissed):
Looks good.

forresttanaka committed:
ENCD-5614-cart-dataset-filters

Change analysis property.

forresttanaka committed:
ENCD-5614-cart-dataset-filters

Ge3t analysis facet working post rebase with dev.

forresttanaka committed:
ENCD-5614-cart-dataset-filters

Comment and query-string ordering update.

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -184,79 +185,116 @@ const displayedFacetFields = [
-/** Facet `field` values for properties from dataset instead of files */
-const datasetFacets = displayedFacetFields.concat(requestedFacetFields).filter((facetField) => facetField.dataset).map((facetField) => facetField.field);
-
+// const requestedFacetFields = displayedFacetFields.filter((field) => !field.calculated).concat([
```
Why is this not removed?

forresttanaka replied:
Oh dang; I forgot to remove it.

zoldello reviewed (dismissed):
Looks good.

forresttanaka commented:
Closing to implement a completely different design.
